### PR TITLE
buildRetryStrategy + CustomRetryStrategyError

### PR DESCRIPTION
**Pull request by timolegros**
## Link to Issue
Closes: #7638

## Description of Changes
- Adds a `buildRetryStrategy` function
- Added types for RetryStrategies
- Added a `CustomRetryStrategyError` class used to define custom retry strategies for edge-cases

## Test Plan
- Ensure this env var is set `ALLOWED_EVENTS=ThreadCreated`
- Start RabbitMQ
- Start the local server
- Start the message relayer with `pnpm workspace commonwealth start-message-relayer`
- Create a thread in any community
  - The message relayer logs should show that a thread created event was successfully published

## Deployment Plan

1. 

## Other Considerations

-

### CODE REVIEW

**Comment by timolegros** [comment]
@rbennettcw lmk what you think of the changes. For RPC errors in the contest policy, the default retry strategy is retry 3 times with 2 second intervals but you can easily call the `buildRetryStrategy` function when subscribing to modify the default retry count or interval for all errors. You can also add additional mapping to handle various errors differently. For example, if you want to retry only once for `ECONNREFUSED` errors and increase retry attempts for all other errors to 5, it would look like this:

```
const connRefusedStrategy = (
  err: Error | InvalidInput | CustomRetryStrategyError,
  topic: BrokerSubscriptions,
  content: any,
  ackOrNackFn: (...args: any[]) => void,
  log: ILogger,
) => {
  if (err.code === 'ECONNREFUSED') {
    ackOrNackFn(err, [{ strategy: 'republish', attempts: 1, defer: 10}, { strategy: 'nack' }]);
    return true;
  }
  return false;
}

const contestWorkerSubRes = await brokerInstance.subscribe(
  BrokerSubscriptions.ContestWorkerPolicy,
  ContestWorker(),
  buildRetryStrategy(connRefusedStrategy, 5)
);
```

File: libs/adapters/src/rabbitmq/RabbitMQAdapter.ts [hunk 1]
```diff
@@ -16,30 +17,63 @@ import { Message } from 'amqplib';
...
+      ackOrNackFn(err, err.recoveryStrategy);
+      return;
+    }
+
+    let res = false;
+    if (customRetryStrategiesMap) {
+      res = customRetryStrategiesMap(err, topic, content, ackOrNackFn, log);
+    }
+
+    if (!res) {
```
Comment (Rotorsoft) [inline libs/adapters/src/rabbitmq/RabbitMQAdapter.ts:61]: Why not making strategies map a mandatory argument? seems like we will always need a strategy (with a sensible default)... then this like would look like:
```
if( ! retryStrategies(err, topic, content, ...)) {
...
}
```

**Review by Rotorsoft — APPROVED** [review]
LGTM , just a small comment

File: libs/adapters/src/rabbitmq/RabbitMQAdapter.ts (hunk 1 above)
Comment (timolegros) [reply libs/adapters/src/rabbitmq/RabbitMQAdapter.ts:61]: Actually, the opposite seems to be true. We almost always don't need a strategy since the default retry X times in Y intervals default covers most cases. For the contest projection handlers for example we only need to modify the default retry strategy to increase X and Y and don't need to add an additional strategy for specific errors.

Comment (Rotorsoft) [reply libs/adapters/src/rabbitmq/RabbitMQAdapter.ts:61]: ok.. I was thinking about an explicit default strategy to make the strategy argument required

File: packages/commonwealth/server/workers/commonwealthConsumer/commonwealthConsumer.ts [hunk 2]
```diff
@@ -70,6 +71,7 @@ export async function setupCommonwealthConsumer(): Promise<void> {
   const contestWorkerSubRes = await brokerInstance.subscribe(
     BrokerSubscriptions.ContestWorkerPolicy,
     ContestWorker(),
+    buildRetryStrategy(undefined, 20_000),
```
Comment (rbennettcw) [inline packages/commonwealth/server/workers/commonwealthConsumer/commonwealthConsumer.ts:74]: Just so I'm understanding correctly– this is using the default config (3 retries 2s delay), but overriding the delay to 20s?

Also, is it possible to use exponential backoff instead of linear? Not strictly required but it would be great for some things like RPC rate limiting.

Comment (timolegros) [reply packages/commonwealth/server/workers/commonwealthConsumer/commonwealthConsumer.ts:74]: Rascal doesn't natively support exponential backoff but what you could do is generate an array of republish strategies which achieves the same thing i.e.:
```
[{ strategy: 'republish', defer: 20_000, attempts: 1}, { strategy: 'republish', defer: 40_000, attempts: 1}, {strategy: 'republish', defer: 120_000, attempts: 1}, ..., { strategy: 'nack' }];
```

Comment (timolegros) [reply packages/commonwealth/server/workers/commonwealthConsumer/commonwealthConsumer.ts:74]: If this is something we want I recommend we merge the PR as is for now and then open a separate ticket for modifying the strategy to exponential backoff.

Comment (timolegros) [reply packages/commonwealth/server/workers/commonwealthConsumer/commonwealthConsumer.ts:74]: And to answer your initial question, yes, `buildRetryStrategy(undefined, 20_000)` uses the default config and just increases the defer interval to 20 seconds.

File: libs/adapters/src/rabbitmq/RabbitMQAdapter.ts (hunk 1 above)
Comment (Rotorsoft) [reply libs/adapters/src/rabbitmq/RabbitMQAdapter.ts:61]: I mean an explicit default argument like 

```
function build(strategy = defaultStrategy) {
...
}
```

**Review by rbennettcw — APPROVED** [review]

File: libs/adapters/src/rabbitmq/RabbitMQAdapter.ts (hunk 1 above)
Comment (rbennettcw) [reply libs/adapters/src/rabbitmq/RabbitMQAdapter.ts:61]: Gonna go ahead and resolve/merge this so we can move forward to see if this will fix the race condition.
